replay-test/run.ts

```diff
@@ -0,0 +1,287 @@
+let gRunPattern: string;
+let gDispatchAddress = "wss://dispatch.replay.io";
+
+for (let i = 2; i < process.argv.length; i++) {
+  switch (process.argv[i]) {
+    case "--node":
+      gNodePath = path.resolve(process.cwd(), process.argv[++i]);
+      break;
+    case "--api-key":
+      gAPIKey = process.argv[++i];
+      break;
+    case "--run-suite":
+      gRunSuite = true;
+      break;
+    case "--run-random-tests":
+      gRunRandomTests = +process.argv[++i];
+      break;
+    case "--run-pattern":
+      gRunPattern = process.argv[++i];
+      break;
+    case "--server":
+      gDispatchAddress = process.argv[++i];
+      break;
+    default:
+      bailout(Usage);
+  }
+}
+
+if (!gNodePath) {
+  bailout("Node path not specified");
+}
+
+if (!gAPIKey) {
+  bailout("API key not specified");
+}
+
+if (!gRunSuite && !gRunRandomTests && !gRunPattern) {
+  bailout("No tests specified");
+}
+
+const gRecordingDirectory = path.join(__dirname, `recordings-${(Math.random() * 1e9) | 0}`);
+
+let gNumFailures = 0;
+
+async function main() {
+  fs.mkdirSync(gRecordingDirectory);
+
+  if (gRunSuite) {
+    await runTestSuite();
+  }
+
+  if (gRunRandomTests) {
+    await runRandomTests(gRunRandomTests);
+  }
+
+  if (gRunPattern) {
+    await runTestsMatchingPattern(gRunPattern);
+  }
+
+  fs.rmSync(gRecordingDirectory, { force: true, recursive: true });
+
+  if (gNumFailures) {
+    console.error(`Had ${gNumFailures} test failures`);
+    process.exit(1);
+  }
+
+  console.log("All tests passed, exiting");
+  process.exit(0);
+}
+
+main();
+
+async function runTestSuite() {
+  for (const { name, allowRecordingError } of TestManifest) {
+    await runSingleTest(path.join(__dirname, "examples", name), allowRecordingError);
+  }
+}
+
+async function runRandomTests(count: number) {
+  const testList = readTests(path.join(__dirname, "..", "test"));
+  for (let i = 0; i < count; i++) {
+    const testPath = pickRandomTest();
+    if (testPath) {
+      await runSingleTest(testPath, true);
+    }
+  }
+
+  function pickRandomTest() {
+    const index = (Math.random() * testList.length) | 0;
+    const testPath = testList[index];
+    if (NodeTestIgnoreList.some(pattern => testPath.includes(pattern))) {
+      return null;
+    }
+    return testPath;
+  }
+}
+
+async function runTestsMatchingPattern(pattern: string) {
+  for (const { name, allowRecordingError } of TestManifest) {
+    const testPath = path.join(__dirname, "examples", name);
+    if (testPath.includes(pattern)) {
+      await runSingleTest(testPath, allowRecordingError);
+    }
+  }
+
+  const testList = readTests(path.join(__dirname, "..", "test"));
+  for (const testPath of testList) {
+    if (testPath.includes(pattern)) {
+      await runSingleTest(testPath, true);
+    }
+  }
+}
+
+// Get the full paths to all JS files in a directory.
+function readTests(directory: string): string[] {
+  const rv: string[] = [];
+  const entries = fs.readdirSync(directory, { withFileTypes: true });
```

Comment:
nit: same as above re: fs/promises and not using the *Sync `fs` APIs. Also not a blocker.